types/rpc-handler.ts

```diff
@@ -60,26 +61,113 @@ export class RPCHandler implements HandlerInterface {
+    this._rpcService = new RPCService(this);
   }
 
   /**
-   * Loops through all RPCs for a given network id and returns a provider with the first successful network.
+   * @NOTICE # This method is intended for read operations only.
+   * Performing write operations with this method is dangerous as it is executed and confirmed by potentially all supported network providers.
+   *
+   * @DEV `quorumThreshold` is a template literal to enforce at the type level that the value is a decimal
+   * between 0 and 1 without the need for external packages or custom classes which burden the user and codebase.
+   *
+   * @DESCRIPTION Similar to `Ethers.FallbackProvider`, this method validates the response from multiple nodes and returns true if the response is consistent across all nodes above the quorum threshold.
+   *
+   * @EXAMPLE
+   * - `consensusCall({ method: "eth_blockNumber", params: [] }, "0.5")`
+   * - `consensusCall({ method: "eth_getTransactionByHash", params: ["0x1234"] }, "0.8")`
    */
-  public async getFirstAvailableRpcProvider() {
-    const rpcList = [...networkRpcs[this._networkId].rpcs];
-    shuffleArray(rpcList);
-    for (const rpc of rpcList) {
+  public async consensusCall<TMethodReturnData = unknown>(requestPayload: RequestPayload, quorumThreshold: `0.${number}`): Promise<TMethodReturnData> {
+    if (this._runtimeRpcs.length === 0) {
+      await this.testRpcPerformance();
+    }
+
+    if (this._runtimeRpcs.length === 1) {
+      throw new Error("Only one RPC available, could not reach consensus");
+    }
+
+    const quorum = Math.ceil(this._runtimeRpcs.length * Number(quorumThreshold));
+    const rpcs = this._runtimeRpcs;
+    const results: PromiseResult[] = [];
+
+    for (const rpc of rpcs) {
       try {
-        const result = await RPCService.makeRpcRequest(rpc.url, this._rpcTimeout, { "Content-Type": "application/json" });
+        const result = await this._rpcService.makeRpcRequest(requestPayload, { rpcUrl: rpc, rpcTimeout: this._rpcTimeout });
         if (result.success) {
-          return new JsonRpcProvider({ url: rpc.url, skipFetchSetup: true }, Number(this._networkId));
-        } else {
-          console.error(`Failed to reach endpoint ${rpc.url}. ${result.error}`);
+          results.push(result);
         }
       } catch (err) {
-        console.error(`Failed to reach endpoint ${rpc.url}. ${err}`);
+        this.log("error", `Failed to reach endpoint ${rpc}.\n ${String(err)}`);
       }
     }
+
+    if (results.length < quorum) {
+      throw new Error(`Failed to reach consensus of ${quorum} with ${results.length} successful RPC responses`);
+    }
+
+    const rpcResults = results.map((res) => res.data?.result);
+
+    const matchingResults = rpcResults.reduce(
+      (acc, val) => {
+        if (!val) return acc;
+        if (typeof val !== "string") {
+          if (val instanceof Error) {
+            val = val.message;
+          } else {
+            val = val.hash;
```

Comment:
Some of the RPC methods (like [eth_getTransactionReceipt](https://docs.metamask.io/services/reference/ethereum/json-rpc-methods/eth_gettransactionreceipt/)) don't have the `hash` property in the response so I don't fully understand  the logic

Reply:
I debated this myself a little.

I went with `JSON.Stringify` at first and then return returning `JSON.parse` but some response data is enormous so I went with `hash` instead.

Other than `JSON.Stringify` I'm not sure how we can match responses with certainty.

Reply:
what about `.hash ?? .transactionHash` I feel that might be better but still a lot of methods would be excluded.

Perhaps the method takes a a 3rd param whih is the prop of the return data that we need to compare and then allow the user to tell us?

Reply:
I've implemented `.hash ?? .transactionHash` for now as that covers `eth_getTransactionReceipt` at least.

A 3rd param sounds scalable across all `eth_` methods, failing that, stringifying the results might be the best bet.

Reply:
`JSON.Stringify` sounds good, we need a generalized solution that covers most of RPC methods, not just 2

Reply:
So the trouble with using the entire response object is that they are not 100% consistent across all providers. So we sort all props and sub props of the response and then stringify but this isn't foolproof because even then providers inject additional data that others don't.

One example I've seen is `author` being present in some `getBlockByNumber` while not in others but I'm not sure how big of an issue this is exactly.

I think this is reasonable for V1 but if it struggles for this reason I think 3rd param might be the way forward.